checker/src/types/helpers.rs

```diff
@@ -0,0 +1,471 @@
+}
+
+// unfolds narrowing
+#[must_use]
+pub fn get_origin(ty: TypeId, types: &TypeStore) -> TypeId {
+	if let Type::Narrowed { from, .. } = types.get_type_by_id(ty) {
+		// Hopefully don't have a nested from
+		*from
+	} else {
+		ty
+	}
+}
+
+/// Temp fix for equality of narrowing stuff
+#[must_use]
+pub fn is_not_of_constant(ty: TypeId, types: &TypeStore) -> bool {
+	if let Type::PartiallyAppliedGenerics(PartiallyAppliedGenerics {
+		on: TypeId::NOT_RESTRICTION,
+		arguments,
+	}) = types.get_type_by_id(ty)
+	{
+		let inner = arguments.get_structure_restriction(TypeId::T_TYPE).unwrap();
+		types.get_type_by_id(inner).is_constant()
+	} else {
+		false
+	}
+}
+
+// TODO narrowed as well
+#[must_use]
+pub fn type_equal(lhs: TypeId, rhs: TypeId, types: &TypeStore) -> bool {
+	if lhs == rhs {
+		true
+	} else if let (Type::Constant(lhs), Type::Constant(rhs)) =
+		(types.get_type_by_id(lhs), types.get_type_by_id(rhs))
+	{
+		lhs == rhs
+	} else {
+		false
+	}
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct AndCondition(pub TypeId);
+
+#[derive(Debug)]
+pub struct OrCase(pub Vec<AndCondition>);
+
+#[must_use]
+pub fn into_conditions(id: TypeId, types: &TypeStore) -> Vec<AndCondition> {
+	let ty = types.get_type_by_id(id);
+
+	// Temp fix
+	if let Type::Constructor(Constructor::BinaryOperator { result, .. }) = ty {
+		return if matches!(*result, TypeId::NUMBER_TYPE | TypeId::STRING_TYPE) {
+			vec![AndCondition(id)]
+		} else {
+			into_conditions(*result, types)
+		};
+	}
+
+	if let Type::And(lhs, rhs) = ty {
+		let mut buf = into_conditions(*lhs, types);
+		buf.append(&mut into_conditions(*rhs, types));
+		buf
+	} else if let Some(backing) = get_constraint_or_alias(id, types) {
+		into_conditions(backing, types)
+	} else {
+		vec![AndCondition(id)]
+	}
+}
+
+#[must_use]
+pub fn into_cases(id: TypeId, types: &TypeStore) -> Vec<OrCase> {
+	let ty = types.get_type_by_id(id);
+	if let Type::Or(lhs, rhs) = ty {
+		let mut buf = into_cases(*lhs, types);
+		buf.append(&mut into_cases(*rhs, types));
+		buf
+	} else if let Some(backing) = get_constraint_or_alias(id, types) {
+		into_cases(backing, types)
+	} else {
+		vec![OrCase(into_conditions(id, types))]
+	}
+}
+
+/// Returns the constraint or base of a constant for a type. Otherwise just return the type
+#[must_use]
+pub fn get_larger_type(on: TypeId, types: &TypeStore) -> TypeId {
+	if let Some(poly_base) = get_constraint(on, types) {
+		poly_base
+	} else if let Type::Constant(cst) = types.get_type_by_id(on) {
+		cst.get_backing_type()
+	} else {
+		on
+	}
+}
+
+#[must_use]
+pub fn get_constraint_or_alias(on: TypeId, types: &TypeStore) -> Option<TypeId> {
+	match types.get_type_by_id(on) {
+		Type::RootPolyType(rpt) => Some(rpt.get_constraint()),
+		Type::Constructor(constr) => Some(constr.get_constraint()),
+		Type::AliasTo { to, parameters: None, .. } => Some(*to),
+		Type::Narrowed { narrowed_to, .. } => Some(*narrowed_to),
+		_ => None,
+	}
+}
+
+#[derive(Debug)]
```

Comment:
Made a bit of a upgrade to @sor4chi s work on #218 because wanted a bit more of an abstraction to catch disjoint issues.